extensions/security-jpa/deployment/src/main/java/io/quarkus/security/jpa/deployment/QuarkusSecurityJpaProcessor.java

```diff
@@ -0,0 +1,370 @@
+            // FIXME: check that it's in the same class hierarchy
+            // FIXME: look up fields with default names? name, username…
+            AnnotationTarget annotatedPassword = getSingleAnnotatedElement(index.getIndex(), DOTNAME_PASSWORD);
+            AnnotationTarget annotatedRoles = getSingleAnnotatedElement(index.getIndex(), DOTNAME_ROLES);
+            Set<String> panacheEntities = collectPanacheEntities(panacheEntityClasses);
+            // collect associated getters if required
+            JpaSecurityDefinition jpaSecurityDefinition = new JpaSecurityDefinition(index.getIndex(),
+                    userDefinitionClass,
+                    isPanache(userDefinitionClass, panacheEntities),
+                    annotatedUsername,
+                    annotatedPassword,
+                    annotatedRoles);
+            AnnotationInstance passAnnotation = jpaSecurityDefinition.password.annotation(DOTNAME_PASSWORD);
+            AnnotationValue passwordType = passAnnotation.value();
+            generateIdentityProvider(index.getIndex(), jpaSecurityDefinition,
+                    passwordType != null ? passwordType.asEnum() : PasswordType.MCF.name(),
+                    beanProducer, panacheEntities);
+            recorder.registerSecurityProvider();
+        }
+    }
+
+    private boolean isPanache(ClassInfo annotatedClass, Set<String> panacheEntities) {
+        return panacheEntities.contains(annotatedClass.name().toString());
+    }
+
+    private Set<String> collectPanacheEntities(List<PanacheEntityClassesBuildItem> panacheEntityClassesBuildItems) {
+        Set<String> modelClasses = new HashSet<>();
+        for (PanacheEntityClassesBuildItem panacheEntityClasses : panacheEntityClassesBuildItems) {
+            modelClasses.addAll(panacheEntityClasses.getEntityClasses());
+        }
+        return modelClasses;
+    }
+
+    private AnnotationTarget getSingleAnnotatedElement(Index index, DotName annotation) {
+        List<AnnotationInstance> annotations = index.getAnnotations(annotation);
+        if (annotations.isEmpty())
+            return null;
+        else if (annotations.size() > 1)
+            throw new RuntimeException("You can only annotate one field or method with @" + annotation);
+        return annotations.get(0).target();
+    }
+
+    private void generateIdentityProvider(Index index, JpaSecurityDefinition jpaSecurityDefinition, String passwordType,
+            BuildProducer<GeneratedBeanBuildItem> beanProducer, Set<String> panacheClasses) {
+        ClassOutput classOutput = new ClassOutput() {
+            @Override
+            public void write(String name, byte[] data) {
+                beanProducer.produce(new GeneratedBeanBuildItem(name, data));
+            }
+
+            @Override
+            public Writer getSourceWriter(String className) {
+                return new OutputStreamWriter(System.err) {
```

Comment:
Is this supposed to be here or was this a debugging thing? If it is not supposed to be there you can use GeneratedBeanGizmoAdaptor

Reply:
I will remove this.

Reply:
Removed